build/platforms.bzl

```diff
@@ -0,0 +1,187 @@
+        "ppc64le",
+        "s390x",
+    ],
+    "windows": [
+        "amd64",
+    ],
+}
+
+# KUBE_CLIENT_PLATFORMS in hack/lib/golang.sh
+CLIENT_PLATFORMS = {
+    "linux": [
+        "386",
+        "amd64",
+        "arm",
+        "arm64",
+        "ppc64le",
+        "s390x",
+    ],
+    "darwin": [
+        "386",
+        "amd64",
+    ],
+    "windows": [
+        "386",
+        "amd64",
+    ],
+}
+
+# KUBE_TEST_PLATFORMS in hack/lib/golang.sh
+TEST_PLATFORMS = {
+    "linux": [
+        "amd64",
+        "arm",
+        "arm64",
+        "s390x",
+        "ppc64le",
+    ],
+    "darwin": [
+        "amd64",
+    ],
+    "windows": [
+        "amd64",
+    ],
+}
+
+# Helper which produces the ALL_PLATFORMS dictionary, composed of the union of
+# CLIENT, NODE, SERVER, and TEST platforms
+def _all_platforms():
+    all_platforms = {}
+    for platforms in [CLIENT_PLATFORMS, NODE_PLATFORMS, SERVER_PLATFORMS, TEST_PLATFORMS]:
+        for os, archs in platforms.items():
+            all_platforms[os] = sets.union(
+                all_platforms.setdefault(os, sets.make()),
+                sets.make(archs),
+            )
+    for os, archs in all_platforms.items():
+        all_platforms[os] = sets.to_list(archs)
+    return all_platforms
+
+ALL_PLATFORMS = _all_platforms()
+
+def go_platform_constraint(os, arch):
+    return "@io_bazel_rules_go//go/platform:%s_%s" % (os, arch)
+
+# Helper to for_platforms which updates the select() dictionary.
+# d is the dictionary being updated.
+# value is the value to set for each item of platforms, which should
+# be a single platform category dictionary (e.g. SERVER_PLATFORMS).
+# only_os selects one of the OSes in platforms.
+def _update_dict_for_platform_category(d, value, platforms, only_os = None):
+    if not value:
+        return
+    for os, arches in platforms.items():
+        if only_os and os != only_os:
+            continue
+        for arch in arches:
+            constraint = go_platform_constraint(os, arch)
+            fmt_args = {"OS": os, "ARCH": arch}
+            if types.is_list(value):
+                # Format all items in the list, and hope there are no duplicates
+                d.setdefault(constraint, []).extend(
+                    [v.format(**fmt_args) for v in value],
+                )
+            else:
+                # Don't overwrite existing value
+                if constraint in d:
+                    fail("duplicate entry for constraint %s", constraint)
+                if types.is_dict(value):
+                    # Format dictionary values only
+                    d[constraint] = {
+                        dict_key: dict_value.format(**fmt_args)
+                        for dict_key, dict_value in value.items()
+                    }
+                else:
+                    # Hopefully this is just a string
+                    d[constraint] = value.format(**fmt_args)
+
+# for_platforms returns a dictionary to be used with select().
```

Comment:
The main thing I do not understand is how `select()` works, which this doesn't actually answer and probably no one aside from you has used before.

Most of the rest is just python, which there is a lot more familiarity about in the community.

IMO a link to bazel's select docs would be helpful: https://docs.bazel.build/versions/master/be/functions.html#select

Reply:
done